KAFKASTORE_TIMEOUT_CONFIG is too low in VMs

Tests are failing when run in VMs because this value is too low for some operations. Here's the stacktrace:

```
Running io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClientTest
[2015-02-03 21:47:55,533] INFO SchemaRegistryConfig values:
        port = 41077
        kafkastore.timeout.ms = 500
        kafkastore.write.max.retries = 5
        debug = true
        request.logger.name = io.confluent.rest-utils.requests
        metrics.sample.window.ms = 30000
        kafkastore.zk.session.timeout.ms = 10000
        kafkastore.topic = _schemas
        kafkastore.write.retry.backoff.ms = 100
        avro.compatibility.level = none
        shutdown.graceful.ms = 1000
        response.mediatype.preferred = [application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, application/json]
        metrics.jmx.prefix = kafka.schema.registry
        host.name = localhost
        metric.reporters = []
        kafkastore.commit.interval.ms = -1
        kafkastore.connection.url = 127.0.0.1:60125
        metrics.num.samples = 2
        response.mediatype.default = application/vnd.schemaregistry.v1+json
        kafkastore.topic.replication.factor = 3
 (io.confluent.kafka.schemaregistry.rest.SchemaRegistryConfig:135)
[2015-02-03 21:47:56,578] INFO Initialized the consumer offset to -1 (io.confluent.kafka.schemaregistry.storage.KafkaStoreReaderThread:80)
[2015-02-03 21:47:56,955] WARN Creating the schema topic _schemas using a replication factor of 1, which is less than the desired one of 3. If this is a production environment, it's crucial to add more brokers and increase the replication factor of the topic. (io.confluent.kafka.schemaregistry.storage.KafkaStore:169)
[2015-02-03 21:47:57,489] INFO [kafka-store-reader-thread-_schemas], Starting  (io.confluent.kafka.schemaregistry.storage.KafkaStoreReaderThread:68)
[2015-02-03 21:47:58,009] ERROR Error starting the schema registry (io.confluent.kafka.schemaregistry.rest.SchemaRegistryRestApplication:57)
io.confluent.kafka.schemaregistry.storage.exceptions.SchemaRegistryException: Error while initializing the datastore
        at io.confluent.kafka.schemaregistry.storage.KafkaSchemaRegistry.init(KafkaSchemaRegistry.java:128)
        at io.confluent.kafka.schemaregistry.rest.SchemaRegistryRestApplication.setupResources(SchemaRegistryRestApplication.java:55)
        at io.confluent.kafka.schemaregistry.rest.SchemaRegistryRestApplication.setupResources(SchemaRegistryRestApplication.java:37)
        at io.confluent.rest.Application.createServer(Application.java:104)
        at io.confluent.kafka.schemaregistry.RestApp.start(RestApp.java:51)
        at io.confluent.kafka.schemaregistry.ClusterTestHarness.setUp(ClusterTestHarness.java:135)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
        at java.lang.reflect.Method.invoke(Method.java:597)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
        at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
        at java.lang.reflect.Method.invoke(Method.java:597)
        at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
        at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
        at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
Caused by: io.confluent.kafka.schemaregistry.storage.exceptions.StoreInitializationException: io.confluent.kafka.schemaregistry.storage.exceptions.StoreException: Can't fetch latest offset of Kafka topic _schemas after 500 ms
        at io.confluent.kafka.schemaregistry.storage.KafkaStore.init(KafkaStore.java:148)
        at io.confluent.kafka.schemaregistry.storage.KafkaSchemaRegistry.init(KafkaSchemaRegistry.java:126)
        ... 35 more
Caused by: io.confluent.kafka.schemaregistry.storage.exceptions.StoreException: Can't fetch latest offset of Kafka topic _schemas after 500 ms
        at io.confluent.kafka.schemaregistry.storage.KafkaStore.getLatestOffsetOfKafkaTopic(KafkaStore.java:347)
        at io.confluent.kafka.schemaregistry.storage.KafkaStore.waitUntilBootstrapCompletes(KafkaStore.java:217)
        at io.confluent.kafka.schemaregistry.storage.KafkaStore.init(KafkaStore.java:146)
        ... 36 more
```

I was going to just submit a patch to increase it, but it looks like this timeout is used in quite a few places. I suspect the issue here is that the tests rely on topic auto creation and in the VM that is taking longer for some reason.

I think the right solution is probably to add a second timeout option which is used in certain situations, like during KafkaStore.init(), which should be more lax since we sometimes expect it to take longer. Thoughts from people who are more familiar with this code?

--- a/core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java
+++ b/core/src/main/java/io/confluent/kafka/schemaregistry/rest/SchemaRegistryConfig.java
@@ -66,6 +66,10 @@ public class SchemaRegistryConfig extends RestConfig {
    */
   public static final String KAFKASTORE_TIMEOUT_CONFIG = "kafkastore.timeout.ms";
   /**
+   * <code>kafkastore.init.timeout.ms</code>
+   */
+  public static final String KAFKASTORE_INIT_TIMEOUT_CONFIG = "kafkastore.init.timeout.ms";
+  /**
    * <code>kafkastore.commit.interval.ms</code>
    */
   public static final String KAFKASTORE_COMMIT_INTERVAL_MS_CONFIG = "kafkastore.commit.interval.ms";
@@ -96,6 +100,9 @@ public class SchemaRegistryConfig extends RestConfig {
   protected static final String KAFKASTORE_WRITE_RETRY_BACKOFF_MS_DOC =
       "The amount of time in milliseconds to wait before attempting to retry a failed write "
       + "to the Kafka store";
+  protected static final String KAFKASTORE_INIT_TIMEOUT_DOC =
+      "The timeout for initialization of the Kafka store, including creation of the Kafka topic "
+      + "that stores schema data.";
   protected static final String KAFKASTORE_TIMEOUT_DOC =
       "The timeout for an operation on the Kafka store";
   protected static final String KAFKASTORE_COMMIT_INTERVAL_MS_DOC =
@@ -112,7 +119,7 @@ public class SchemaRegistryConfig extends RestConfig {
   private static final ConfigDef config;
   static {
     config = baseConfigDef()
-        .defineOverride(PORT_CONFIG, ConfigDef.Type.INT, SCHEMAREGISTRY_PORT_DEFAULT, 
+        .defineOverride(PORT_CONFIG, ConfigDef.Type.INT, SCHEMAREGISTRY_PORT_DEFAULT,
                         ConfigDef.Importance.LOW, PORT_CONFIG_DOC)
         .defineOverride(RESPONSE_MEDIATYPE_PREFERRED_CONFIG, ConfigDef.Type.LIST,
                         io.confluent.kafka.schemaregistry.client.rest.Versions.PREFERRED_RESPONSE_TYPES,
@@ -137,6 +144,8 @@ public class SchemaRegistryConfig extends RestConfig {
         .define(KAFKASTORE_WRITE_RETRY_BACKOFF_MS_CONFIG, ConfigDef.Type.INT,
                 DEFAULT_KAFKASTORE_WRITE_RETRY_BACKOFF_MS, ConfigDef.Importance.MEDIUM,
                 KAFKASTORE_WRITE_RETRY_BACKOFF_MS_DOC)
+        .define(KAFKASTORE_INIT_TIMEOUT_CONFIG, ConfigDef.Type.INT, 5000, atLeast(0),
+                ConfigDef.Importance.MEDIUM, KAFKASTORE_INIT_TIMEOUT_DOC)
         .define(KAFKASTORE_TIMEOUT_CONFIG, ConfigDef.Type.INT, 500, atLeast(0),
                 ConfigDef.Importance.MEDIUM, KAFKASTORE_TIMEOUT_DOC)
         .define(KAFKASTORE_COMMIT_INTERVAL_MS_CONFIG, ConfigDef.Type.INT,

--- a/core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java
+++ b/core/src/main/java/io/confluent/kafka/schemaregistry/storage/KafkaStore.java
@@ -73,6 +73,7 @@ public class KafkaStore<K, V> implements Store<K, V> {
   private final Serializer<K, V> serializer;
   private final Store<K, V> localStore;
   private final AtomicBoolean initialized = new AtomicBoolean(false);
+  private final int initTimeout;
   private final int timeout;
   private final Seq<Broker> brokerSeq;
   private final ZkClient zkClient;
@@ -95,6 +96,7 @@ public class KafkaStore<K, V> implements Store<K, V> {
     this.groupId = String.format("schema-registry-%s-%d",
                                  config.getString(SchemaRegistryConfig.HOST_NAME_CONFIG),
                                  config.getInt(SchemaRegistryConfig.PORT_CONFIG));
+    initTimeout = config.getInt(SchemaRegistryConfig.KAFKASTORE_INIT_TIMEOUT_CONFIG);
     timeout = config.getInt(SchemaRegistryConfig.KAFKASTORE_TIMEOUT_CONFIG);
     this.storeUpdateHandler = storeUpdateHandler;
     this.serializer = serializer;
@@ -144,7 +146,7 @@ public class KafkaStore<K, V> implements Store<K, V> {
     kafkaTopicReader.start();
 
     try {
-      waitUntilBootstrapCompletes();
+      waitUntilBootstrapCompletes(initTimeout);
     } catch (StoreException e) {
       throw new StoreInitializationException(e);
     }
@@ -215,9 +217,15 @@ public class KafkaStore<K, V> implements Store<K, V> {
    * Wait until the KafkaStore catches up to the last message in the Kafka topic.
    */
   public void waitUntilBootstrapCompletes() throws StoreException {
-    long offsetOfLastMessage = getLatestOffsetOfKafkaTopic(timeout) - 1;
+    waitUntilBootstrapCompletes(timeout);
+  }
+  /**
+   * Wait until the KafkaStore catches up to the last message in the Kafka topic.
+   */
+  private void waitUntilBootstrapCompletes(int timeoutMs) throws StoreException {
+    long offsetOfLastMessage = getLatestOffsetOfKafkaTopic(timeoutMs) - 1;
     log.info("Wait to catch up until the offset of the last message at " + offsetOfLastMessage);
-    kafkaTopicReader.waitUntilOffset(offsetOfLastMessage, timeout, TimeUnit.MILLISECONDS);
+    kafkaTopicReader.waitUntilOffset(offsetOfLastMessage, timeoutMs, TimeUnit.MILLISECONDS);
     log.debug("Reached offset at " + offsetOfLastMessage);
   }
 

